### PR TITLE
bpo-47015: Update test_os from asyncore to asyncio

### PR DESCRIPTION
**Pull request by arhadthedev**
[PEP 594 – Removing dead batteries from the standard library](https://peps.python.org/pep-0594/) removes asyncore and asynchat in 3.12 with the following note:

> The `asyncore` module is also used in stdlib tests. The tests for `ftplib`, `logging`, `smptd`, `smtplib`, and `ssl` are partly based on asyncore. These tests must be updated to use asyncio or threading.

However, the citation misses the tests for `os`. So this PR ports `test_os.TestSendfile` as the only class that uses asyncore to create a dummy sendfile server and communicate with it.

To simplify the review I broke all changes into a few commits to get readable diffs:

1. Preparation:
    1. Change inheritance of the test from `TestCase` to `IsolatedAsyncioTestCase`
    1. Make `os.sendfile` async-friendly by moving it into a worker thread to not block an event loop of IsolatedAsyncioTestCase. As a result, `sendfile_wrapper` that called it is also made async
    1. Explicitly specify a global event loop policy because an implicit initialization
 by `IsolatedAsyncioTestCase` is counted as a test failure (this commit is actually the last)
1. Replace `SendfileTestServer` with `asyncio.start_server`. The client side is left mostly untouched because `os.sendfile` works directly with a low-level socket ignoring asyncio message loop anyway. "Mostly" because `connect` is delegated to an event loop to not block the server in the same thread
1. Cleanup:
    1. Convert server-side buffer population from a loop into a list comprehension to be more Pythonic
    1. Remove redundant waiting of the client for `220 ready`. Instead, we use a guarantee that `asyncio.loop.sock_connect` yields only after establishing a connection
    1. Remove `SendfileTestServer` with related asyncio imports
    1. Remove threading leftovers (`asyncio.to_thread` maintains its own threading machinery)

https://bugs.python.org/issue47015

### CODE REVIEW

**Comment by arhadthedev** [comment]
Note to myself: despite `asyncio.start_server` promises to be IPv4-first if `family` is undefined:

> https://github.com/python/cpython/blob/9523c0d84f351a610dc651b234461eb015fa3b82/Lib/asyncio/base_events.py#L976
>
> calls
> https://github.com/python/cpython/blob/9523c0d84f351a610dc651b234461eb015fa3b82/Lib/asyncio/base_events.py#L1034-L1036
>
> that calls
>
> https://github.com/python/cpython/blob/9523c0d84f351a610dc651b234461eb015fa3b82/Lib/asyncio/base_events.py#L1400
>
> that has the following lines:
>
> https://github.com/python/cpython/blob/9523c0d84f351a610dc651b234461eb015fa3b82/Lib/asyncio/base_events.py#L129-L134

, it creates an IPv6-only listening socket if possible. As a result, not only `server.sockets[0].getsockname()` returns an IPv6 address, the whole server refuses to accept IPv4 client connections. It took me two force-pushes to find out, so my apologises about flukes in the notifications.

Now, I'm sorting out what and why leaks all connected sockets (as Address sanitizer [reports](https://github.com/python/cpython/runs/5559966553?check_suite_focus=true#step:13:311)) failing all non-Windows runners:

> /home/runner/work/cpython/cpython/Lib/unittest/async_case.py:134: ResourceWarning: unclosed <socket.socket fd=8, family=2, type=1, proto=6, laddr=('127.0.0.1', 32881), raddr=('127.0.0.1', 57190)>
>
> /home/runner/work/cpython/cpython/Lib/unittest/async_case.py:134: ResourceWarning: unclosed <socket.socket fd=10, family=2, type=1, proto=6, laddr=('127.0.0.1', 36545), raddr=('127.0.0.1', 42800)>
>
> /home/runner/work/cpython/cpython/Lib/unittest/async_case.py:134: ResourceWarning: unclosed <socket.socket fd=12, family=2, type=1, proto=6, laddr=('127.0.0.1', 35363), raddr=('127.0.0.1', 49040)>
>
> /home/runner/work/cpython/cpython/Lib/unittest/async_case.py:134: ResourceWarning: unclosed <socket.socket fd=14, family=2, type=1, proto=6, laddr=('127.0.0.1', 39391), raddr=('127.0.0.1', 41928)>
>
> /home/runner/work/cpython/cpython/Lib/unittest/async_case.py:134: ResourceWarning: unclosed <socket.socket fd=16, family=2, type=1, proto=6, laddr=('127.0.0.1', 42821), raddr=('127.0.0.1', 42452)>

**Comment by arhadthedev** [comment]
The failure was a combination of client-side blocking wait of a server and a missing asyncio loop policy. Both problems are fixed.

File: Lib/test/test_os.py [hunk 1]
```diff
@@ -3328,40 +3238,52 @@ class TestSendfile(unittest.TestCase):
...
+
+    async def handle_new_client(self, reader, writer):
+        self.server_buffer = b''.join([x async for x in self.chunks(reader)])
+        writer.close()
+        self.server.close()  # The test server processes a single client only
+
+    async def asyncSetUp(self):
+        self.server_buffer = b''
+        cb = lambda reader, writer: self.handle_new_client(reader, writer)
+        self.server = await asyncio.start_server(cb, socket_helper.HOSTv4)
```
Comment (merwok) [inline Lib/test/test_os.py]: It the callback is a function with `reader, writer` params, why not pass `self.handle_new_client` directly without the intervening lambda?

Comment (arhadthedev) [reply Lib/test/test_os.py]: Thank you, I didn't know that Python autocaptures `self` when obtaining a reference to a method so I've done it manually in the lambda first. I've thought that, like in C++ or JavaScript, just getting a method reference without calling it right there will require to specify a call target once more.

Comment (merwok) [reply Lib/test/test_os.py]: It’s the beauty of descriptors!

File: Lib/test/test_os.py [hunk 2]
```diff
@@ -70,6 +64,8 @@
 from test.support.os_helper import FakePath
 
 
+asyncio.set_event_loop_policy(asyncio.DefaultEventLoopPolicy())
```
Comment (asvetlov) [inline Lib/test/test_os.py]: Why is this line needed?

File: Lib/test/test_os.py [hunk 3]
```diff
@@ -70,6 +64,8 @@
...
+
```
Comment (asvetlov) [inline Lib/test/test_os.py]: Please add the following to prevent a warning about unexpected environment modification.
```
def tearDownModule():
    asyncio.set_event_loop_policy(None)
```

File: Lib/test/test_os.py [hunk 4]
```diff
@@ -3328,40 +3238,52 @@ class TestSendfile(unittest.TestCase):
...
+        self.server = await asyncio.start_server(self.handle_new_client,
+                                                 socket_helper.HOSTv4)
         self.client = socket.socket()
-        self.client.connect((self.server.host, self.server.port))
-        self.client.settimeout(1)
-        # synchronize by waiting for "220 ready" response
-        self.client.recv(1024)
+        self.client.setblocking(False)
+        l = asyncio.get_running_loop()
+        await l.sock_connect(self.client, self.server.sockets[0].getsockname())
```
Comment (asvetlov) [inline Lib/test/test_os.py]: Don't hesitate to name a loop as the loop :)
```suggestion
        loop = asyncio.get_running_loop()
        await loop.sock_connect(self.client, self.server.sockets[0].getsockname())
```

**Review by asvetlov — CHANGES_REQUESTED** [review]

File: Lib/test/test_os.py (hunk 3 above)
Comment (arhadthedev) [reply Lib/test/test_os.py]: Thank you, changed.

File: Lib/test/test_os.py (hunk 4 above)
Comment (arhadthedev) [reply Lib/test/test_os.py]: The short name allowed to keep a 79 character limit. Though, on second thoughts, I took out a server address instead:

```diff
-        l = asyncio.get_running_loop()
-        await l.sock_connect(self.client, self.server.sockets[0].getsockname())
+        server_name = self.server.sockets[0].getsockname()
+        await asyncio.get_running_loop().sock_connect(self.client, server_name) 
```

Comment (arhadthedev) [reply Lib/test/test_os.py]: On *the third thoughts*, I moved `server_name = self.server.sockets[0].getsockname()` right under server creation. Without it, reading code as a prose becomes harder:

> Start the server. For a client, create a socket, set it to be nonblocking, get a server name that is a name of its listening port, connect to the server.

instead of:

> Start the server, its name is a name of its listening port. For a client, create a socket, set it to be nonblocking, connect to the server.

File: Lib/test/test_os.py [hunk 5]
```diff
@@ -101,6 +95,10 @@ def create_file(filename, content=b'content'):
                                        'on AIX, splice() only accepts sockets')
 
 
+def tearDownModule():
+    asyncio.set_event_loop_policy(None)
```
Comment (merwok) [inline Lib/test/test_os.py:99]: The line that set a loop policy is not present any more, so this cleanup step doesn’t seem to apply anymore.

Comment (arhadthedev) [reply Lib/test/test_os.py:99]: https://github.com/python/cpython/runs/5604663664?check_suite_focus=true

```
901: == Tests result: ENV CHANGED ==

```

Comment (arhadthedev) [reply Lib/test/test_os.py:99]: If I understand [a failed pre-policy run](https://github.com/python/cpython/runs/5604663664?check_suite_focus=true) correctly, the policy must either be manually set to something before any case is runned or be returned back to `None` before the suit exits. I've chosen the first variant, Andrew proposed to switch to the second.

```plain
0:04:29 load avg: 3.56 [300/433/1] test_os failed (env changed) -- running: test_asyncio (1 min 10 sec), test_concurrent_futures (1 min 57 sec)

[a bunch of oks and skips follow]

[...]

423 tests OK.

10 slowest tests:
- test_tools: 4 min 44 sec
- test_concurrent_futures: 3 min 30 sec
- test_peg_generator: 2 min 57 sec
- test_multiprocessing_spawn: 2 min 27 sec
- test_gdb: 1 min 51 sec
- test_asyncio: 1 min 50 sec
- test_multiprocessing_forkserver: 1 min 38 sec
- test_multiprocessing_fork: 1 min 22 sec
- test_regrtest: 1 min 12 sec
- test_statistics: 49.1 sec

1 test altered the execution environment:
    test_os

9 tests skipped:
    test_devpoll test_ioctl test_kqueue test_msilib test_startfile
    test_winconsoleio test_winreg test_winsound test_zipfile64

Total duration: 11 min 50 sec
Tests result: ENV CHANGED
make: *** [Makefile:1795: buildbottest] Error 3
Error: Process completed with exit code 2.
```

**Comment by asvetlov** [comment]
LGTM. Let's wait for the buildbots fleet to finish to make sure that everything works smoothly.
Network code can lead to subtle test problems.

**Comment by arhadthedev** [comment]
@asvetlov I fixed a forgotten `self` found by AMD64 FreeBSD.

The second runner, ARM64 macOS, failed on `test_asyncio` that is unrelated:

```
======================================================================
ERROR: test_get_cancelled (test.test_asyncio.test_queues.QueueGetTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/buildbot/buildarea/pull_request.pablogsal-macos-m1.macos-with-brew/build/Lib/asyncio/tasks.py", line 490, in wait_for
    return fut.result()
           ^^^^^^^^^^^^
  File "/Users/buildbot/buildarea/pull_request.pablogsal-macos-m1.macos-with-brew/build/Lib/asyncio/queues.py", line 158, in get
    await getter
    ^^^^^^^^^^^^
asyncio.exceptions.CancelledError
```

> Network code can lead to subtle test problems.

Considering that the failed `test_flags` was marked as `@requires_headers_trailers` (`not sys.platform.startswith("linux") and not sys.platform.startswith("solaris") and not sys.platform.startswith("sunos")`), GitHub runners had no chance.

**Comment by arhadthedev** [comment]
GitHub Ubuntu runner failed with `Error: Cache service responded with 400 during upload chunk`; I reopened the PR to retrigger the workflow.

**Comment by asvetlov** [comment]
I see. Test failures seem unrelated.
Let's run the buildbot fleet again.

**Comment by asvetlov** [comment]
Merged. Thanks!
